internal/ingest/ingest.go

```diff
@@ -135,7 +136,13 @@ func (i *legIngester) Sync(ctx context.Context, p peer.ID, opts ...ingestion.Syn
 }
 
 func (i *legIngester) getLatestAdvID(ctx context.Context, p peer.ID) (cid.Cid, error) {
-	res, err := i.client.GetLatestAdv(ctx, p)
+	client, err := i.newClient(ctx, i.host, p)
```

Comment:
What happens if two routines call a function for the same peer simultaneously? We will instantiate two different clients instance to the same peer and attempt to establish a stream from scratch? I don't think this is a problem because IIRC the libp2p host will check if it is already connected to the peer and leverage the existing stream. 

What I am not that sure is what would happen if we try to create a new stream while a previous one is still being negotiated. This was the reason for the message sender. Anyway, I think this can be easily tested.

Reply:
I think this is a fairly normal occurrence in libp2p, when two goroutines try to talk to the same peer using the same protocol ID.  Either the existing stream gets reused or a new stream is created.

If we try to create a new stream while a previous one is still being negotiated, then either libp2p will wait until the first is created, or it will create a new stream to the same host with the same protocol ID.  I would guess that it would create a new stream, so that either stream could be reset independently.

This has got to be a common occurrence because different goroutines sharing the same host will try to make connections to some peer/proto.  The different goroutines should not have to coordinate which host they connect to.

This should be documented behavior, but lets test anyway.

Reply:
I ran a test with a few dozen clients connecting to the same server in parallel and nothing broke, so I am happy with this :)